Create new check RecordComponentName

From #8764:

```
➜  src /usr/lib/jvm/java-14-openjdk/bin/javac --enable-preview --source 14 TestClass.java                               
Note: TestClass.java uses preview language features.
Note: Recompile with -Xlint:preview for details.
➜  src cat config.xml                                                                                                   
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
        "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
        "http://www.puppycrawl.com/dtds/configuration_1_3.dtd">
<module name="Checker">
    <property name="charset" value="UTF-8"/>
    <module name="TreeWalker">
        <module name="LocalVariableName">
            <property name="format" value="(?i)^(?!(record|yield|var)$).+$"/>
        </module>
        <module name="TypeName">
            <property name="format" value="(?i)^(?!(record|yield|var)$).+$"/>
        </module>
    </module>
</module>%                                                                                                                     ➜  src cat TestClass.java                                                                                               
public class TestClass {
    public static void main(String... args) {
        var var = 4; // violation, "var" should not be used as an identifier.

        int record = 15; // violation, "record" should not be used as an identifier.

        String yield = "yield"; // violation, "yield" should not be used as an identifier.

        record Record // violation, "Record" should not be used as an idenitifier.
                (Record record) { // should be violation, "record" should not be used as an identifier.
        }

        String yieldString = "yieldString"; // ok, part of another word
        record MyRecord(){} // ok, part of another word
        var variable = 2; // ok, part of another word
    }
}
➜  src java -jar /home/nick/IdeaProjects/checkstyle/target/checkstyle-8.36-SNAPSHOT-all.jar -c config.xml TestClass.java\

Starting audit...
[ERROR] /home/nick/Desktop/full-record-grammar/check-updates/src/TestClass.java:3:13: Name 'var' must match pattern '(?i)^(?!(record|yield|var)$).+$'. [LocalVariableName]
[ERROR] /home/nick/Desktop/full-record-grammar/check-updates/src/TestClass.java:5:13: Name 'record' must match pattern '(?i)^(?!(record|yield|var)$).+$'. [LocalVariableName]
[ERROR] /home/nick/Desktop/full-record-grammar/check-updates/src/TestClass.java:7:16: Name 'yield' must match pattern '(?i)^(?!(record|yield|var)$).+$'. [LocalVariableName]
[ERROR] /home/nick/Desktop/full-record-grammar/check-updates/src/TestClass.java:9:16: Name 'Record' must match pattern '(?i)^(?!(record|yield|var)$).+$'. [TypeName]
Audit done.
Checkstyle ends with 4 errors.

```
I would expect there to be a check that allows users to enfore naming conventions on record components; there should be an ability to specify a `format` property, similar to `MemberNameCheck` (after all, record components are essentially member variables).  Simply adding support for record components to `MemberNameCheck` is not a good solution, since there are many other properties in `MemberNameCheck` that do not apply here. 

Do we need another new check, `RecordComponentName`?

We have LambdaParameterName, MemberName, and  ParameterName.  Since record components are basically "members", we should make it possible for users to enforce naming conventions on them.

--- a/.ci/jsoref-spellchecker/whitelist.words
+++ b/.ci/jsoref-spellchecker/whitelist.words
@@ -1130,6 +1130,7 @@ README
 Readonly
 rebased
 rebasing
+recordcomponentname
 recordcomponentnumber
 recordtypeparametername
 redundantimport

--- a/config/checkstyle_checks.xml
+++ b/config/checkstyle_checks.xml
@@ -690,6 +690,7 @@
     <module name="AbstractClassName"/>
     <module name="ClassTypeParameterName"/>
     <module name="RecordTypeParameterName"/>
+    <module name="RecordComponentName"/>
     <module name="ConstantName"/>
     <module name="InterfaceTypeParameterName"/>
     <module name="LocalFinalVariableName"/>

--- a/src/it/java/com/google/checkstyle/test/chapter5naming/rule526parameternames/RecordComponentNameTest.java
+++ b/src/it/java/com/google/checkstyle/test/chapter5naming/rule526parameternames/RecordComponentNameTest.java
@@ -1,0 +1,54 @@
+////////////////////////////////////////////////////////////////////////////////
+// checkstyle: Checks Java source code for adherence to a set of rules.
+// Copyright (C) 2001-2021 the original author or authors.
+//
+// This library is free software; you can redistribute it and/or
+// modify it under the terms of the GNU Lesser General Public
+// License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+//
+// This library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+// Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public
+// License along with this library; if not, write to the Free Software
+// Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
+////////////////////////////////////////////////////////////////////////////////
+
+package com.google.checkstyle.test.chapter5naming.rule526parameternames;
+
+import java.util.Map;
+
+import org.junit.jupiter.api.Test;
+
+import com.google.checkstyle.test.base.AbstractGoogleModuleTestSupport;
+import com.puppycrawl.tools.checkstyle.api.Configuration;
+
+public class RecordComponentNameTest extends AbstractGoogleModuleTestSupport {
+
+    private static final String MSG_KEY = "name.invalidPattern";
+
+    @Override
+    protected String getPackageLocation() {
+        return "com/google/checkstyle/test/chapter5naming/rule526parameternames";
+    }
+
+    @Test
+    public void testGeneralParameterName() throws Exception {
+        final Configuration config = getModuleConfig("RecordComponentName");
+        final String format = config.getAttribute("format");
+        final Map<String, String> messages = config.getMessages();
+        final String[] expected = {
+            "8:47: " + getCheckMessage(messages, MSG_KEY, "_componentName", format),
+            "12:40: " + getCheckMessage(messages, MSG_KEY, "Capital", format),
+        };
+
+        final String filePath = getNonCompilablePath("InputRecordComponentName.java");
+
+        final Integer[] warnList = getLinesWithWarn(filePath);
+        verify(config, filePath, expected, warnList);
+    }
+
+}

--- a/src/it/java/org/checkstyle/suppressionxpathfilter/XpathRegressionRecordComponentNameTest.java
+++ b/src/it/java/org/checkstyle/suppressionxpathfilter/XpathRegressionRecordComponentNameTest.java
@@ -1,0 +1,88 @@
+////////////////////////////////////////////////////////////////////////////////
+// checkstyle: Checks Java source code for adherence to a set of rules.
+// Copyright (C) 2001-2021 the original author or authors.
+//
+// This library is free software; you can redistribute it and/or
+// modify it under the terms of the GNU Lesser General Public
+// License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+//
+// This library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+// Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public
+// License along with this library; if not, write to the Free Software
+// Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
+////////////////////////////////////////////////////////////////////////////////
+
+package org.checkstyle.suppressionxpathfilter;
+
+import java.io.File;
+import java.util.Collections;
+import java.util.List;
+
+import org.junit.jupiter.api.Test;
+
+import com.puppycrawl.tools.checkstyle.DefaultConfiguration;
+import com.puppycrawl.tools.checkstyle.checks.naming.AbstractNameCheck;
+import com.puppycrawl.tools.checkstyle.checks.naming.RecordComponentNameCheck;
+
+public class XpathRegressionRecordComponentNameTest extends AbstractXpathTestSupport {
+
+    private final String checkName = RecordComponentNameCheck.class.getSimpleName();
+
+    @Override
+    protected String getCheckName() {
+        return checkName;
+    }
+
+    @Test
+    public void testOne() throws Exception {
+        final File fileToProcess = new File(getNonCompilablePath(
+                "SuppressionXpathRecordComponentName1.java"));
+
+        final DefaultConfiguration moduleConfig =
+                createModuleConfig(RecordComponentNameCheck.class);
+
+        final String[] expectedViolation = {
+            "6:56: " + getCheckMessage(RecordComponentNameCheck.class,
+                AbstractNameCheck.MSG_INVALID_PATTERN,
+                    "_value", "^[a-z][a-zA-Z0-9]*$"),
+        };
+
+        final List<String> expectedXpathQueries = Collections.singletonList(
+            "/RECORD_DEF[./IDENT[@text='SuppressionXpathRecordComponentName1']]"
+                + "/RECORD_COMPONENTS/RECORD_COMPONENT_DEF/IDENT[@text='_value']"
+        );
+
+        runVerifications(moduleConfig, fileToProcess, expectedViolation,
+                expectedXpathQueries);
+    }
+
+    @Test
+    public void testTwo() throws Exception {
+        final File fileToProcess = new File(getNonCompilablePath(
+                "SuppressionXpathRecordComponentName2.java"));
+
+        final DefaultConfiguration moduleConfig =
+                createModuleConfig(RecordComponentNameCheck.class);
+        moduleConfig.addAttribute("format", "^_[a-z][a-zA-Z0-9]*$");
+
+        final String[] expectedViolation = {
+            "9:32: " + getCheckMessage(RecordComponentNameCheck.class,
+                AbstractNameCheck.MSG_INVALID_PATTERN,
+                    "otherValue", "^_[a-z][a-zA-Z0-9]*$"),
+        };
+
+        final List<String> expectedXpathQueries = Collections.singletonList(
+            "/CLASS_DEF[./IDENT[@text='SuppressionXpathRecordComponentName2']]/OBJBLOCK"
+                + "/RECORD_DEF[./IDENT[@text='MyRecord']]"
+                + "/RECORD_COMPONENTS/RECORD_COMPONENT_DEF/IDENT[@text='otherValue']"
+        );
+
+        runVerifications(moduleConfig, fileToProcess, expectedViolation,
+                expectedXpathQueries);
+    }
+}

--- a/src/it/resources-noncompilable/com/google/checkstyle/test/chapter5naming/rule526parameternames/InputRecordComponentName.java
+++ b/src/it/resources-noncompilable/com/google/checkstyle/test/chapter5naming/rule526parameternames/InputRecordComponentName.java
@@ -1,0 +1,13 @@
+//non-compiled with javac: Compilable with Java14
+package com.puppycrawl.tools.checkstyle.checks.naming.recordcomponentname;
+
+/* Config:
+ * format = "^[a-z]([a-z0-9][a-zA-Z0-9]*)?$"
+ *
+ */
+public record InputRecordComponentName<t>(int _componentName, // warn
+                                          String componentName2) {
+}
+
+record InputRecordComponentName<t>(int Capital) { // warn
+}

--- a/src/it/resources-noncompilable/org/checkstyle/suppressionxpathfilter/recordcomponentname/SuppressionXpathRecordComponentName1.java
+++ b/src/it/resources-noncompilable/org/checkstyle/suppressionxpathfilter/recordcomponentname/SuppressionXpathRecordComponentName1.java
@@ -1,0 +1,8 @@
+//non-compiled with javac: Compilable with Java14
+package com.puppycrawl.tools.checkstyle.checks.sizes.recordcomponentname;
+
+/* Config: default
+ */
+public record SuppressionXpathRecordComponentName1(int _value) { // warn
+
+}

--- a/src/it/resources-noncompilable/org/checkstyle/suppressionxpathfilter/recordcomponentname/SuppressionXpathRecordComponentName2.java
+++ b/src/it/resources-noncompilable/org/checkstyle/suppressionxpathfilter/recordcomponentname/SuppressionXpathRecordComponentName2.java
@@ -1,0 +1,10 @@
+//non-compiled with javac: Compilable with Java14
+package com.puppycrawl.tools.checkstyle.checks.sizes.recordcomponentname;
+
+/* Config:
+ * format = ^[a-z][a-zA-Z0-9]*$
+ */
+public class SuppressionXpathRecordComponentName2 {
+    public record MyRecord(int _underscoreValue, // ok
+                           int otherValue) { } // warn
+}

--- a/src/main/java/com/puppycrawl/tools/checkstyle/PackageObjectFactory.java
+++ b/src/main/java/com/puppycrawl/tools/checkstyle/PackageObjectFactory.java
@@ -731,8 +731,6 @@ public class PackageObjectFactory implements ModuleFactory {
                 BASE_PACKAGE + ".checks.naming.CatchParameterNameCheck");
         NAME_TO_FULL_MODULE_NAME.put("ClassTypeParameterNameCheck",
                 BASE_PACKAGE + ".checks.naming.ClassTypeParameterNameCheck");
-        NAME_TO_FULL_MODULE_NAME.put("RecordTypeParameterNameCheck",
-                BASE_PACKAGE + ".checks.naming.RecordTypeParameterNameCheck");
         NAME_TO_FULL_MODULE_NAME.put("ConstantNameCheck",
                 BASE_PACKAGE + ".checks.naming.ConstantNameCheck");
         NAME_TO_FULL_MODULE_NAME.put("InterfaceTypeParameterNameCheck",
@@ -753,6 +751,10 @@ public class PackageObjectFactory implements ModuleFactory {
                 BASE_PACKAGE + ".checks.naming.PackageNameCheck");
         NAME_TO_FULL_MODULE_NAME.put("ParameterNameCheck",
                 BASE_PACKAGE + ".checks.naming.ParameterNameCheck");
+        NAME_TO_FULL_MODULE_NAME.put("RecordComponentNameCheck",
+                BASE_PACKAGE + ".checks.naming.RecordComponentNameCheck");
+        NAME_TO_FULL_MODULE_NAME.put("RecordTypeParameterNameCheck",
+                BASE_PACKAGE + ".checks.naming.RecordTypeParameterNameCheck");
         NAME_TO_FULL_MODULE_NAME.put("StaticVariableNameCheck",
                 BASE_PACKAGE + ".checks.naming.StaticVariableNameCheck");
         NAME_TO_FULL_MODULE_NAME.put("TypeNameCheck",

--- a/src/main/java/com/puppycrawl/tools/checkstyle/checks/naming/RecordComponentNameCheck.java
+++ b/src/main/java/com/puppycrawl/tools/checkstyle/checks/naming/RecordComponentNameCheck.java
@@ -1,0 +1,109 @@
+////////////////////////////////////////////////////////////////////////////////
+// checkstyle: Checks Java source code for adherence to a set of rules.
+// Copyright (C) 2001-2021 the original author or authors.
+//
+// This library is free software; you can redistribute it and/or
+// modify it under the terms of the GNU Lesser General Public
+// License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+//
+// This library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+// Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public
+// License along with this library; if not, write to the Free Software
+// Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
+////////////////////////////////////////////////////////////////////////////////
+
+package com.puppycrawl.tools.checkstyle.checks.naming;
+
+import com.puppycrawl.tools.checkstyle.api.DetailAST;
+import com.puppycrawl.tools.checkstyle.api.TokenTypes;
+
+/**
+ * <p>
+ * Checks that record component names conform to a specified pattern.
+ * </p>
+ * <ul>
+ * <li>
+ * Property {@code format} - Specifies valid identifiers.
+ * Type is {@code java.util.regex.Pattern}.
+ * Default value is {@code "^[a-z][a-zA-Z0-9]*$"}.
+ * </li>
+ * </ul>
+ * <p>
+ * To configure the check:
+ * </p>
+ * <pre>
+ * &lt;module name="RecordComponentName"/&gt;
+ * </pre>
+ * <p>Example:</p>
+ * <pre>
+ * record MyRecord1(String value, int otherComponentName) {} // OK
+ * record MyRecord2(String... Values) {} // violation, the record component name
+ *                                     // should match the regular expression "^[a-z][a-zA-Z0-9]*$"
+ * record MyRecord3(double my_number) {} // violation, the record component name
+ *                                     // should match the regular expression "^[a-z][a-zA-Z0-9]*$"
+ * </pre>
+ * <p>
+ * An example of how to configure the check for names that are only letters in lowercase:
+ * </p>
+ * <p>Configuration:</p>
+ * <pre>
+ * &lt;module name="RecordComponentName"&gt;
+ *   &lt;property name="format" value="^[a-z]+$"/&gt;
+ * &lt;/module&gt;
+ * </pre>
+ * <p>Example:</p>
+ * <pre>
+ * record MyRecord1(String value, int other) {} // OK
+ * record MyRecord2(String... strings) {} // OK
+ * record MyRecord3(double myNumber) {} // violation, the record component name
+ *                              // should match the regular expression "^[a-z]+$"
+ * </pre>
+ * <p>
+ * Parent is {@code com.puppycrawl.tools.checkstyle.TreeWalker}
+ * </p>
+ * <p>
+ * Violation Message Keys:
+ * </p>
+ * <ul>
+ * <li>
+ * {@code name.invalidPattern}
+ * </li>
+ * </ul>
+ *
+ * @since 8.40
+ */
+public class RecordComponentNameCheck extends AbstractNameCheck {
+
+    /** Creates a new {@code RecordComponentNameCheck} instance. */
+    public RecordComponentNameCheck() {
+        super("^[a-z][a-zA-Z0-9]*$");
+    }
+
+    @Override
+    public int[] getDefaultTokens() {
+        return getRequiredTokens();
+    }
+
+    @Override
+    public final int[] getAcceptableTokens() {
+        return getRequiredTokens();
+    }
+
+    @Override
+    public int[] getRequiredTokens() {
+        return new int[] {
+            TokenTypes.RECORD_COMPONENT_DEF,
+        };
+    }
+
+    @Override
+    protected final boolean mustCheckName(DetailAST ast) {
+        return true;
+    }
+
+}

--- a/src/main/resources/com/puppycrawl/tools/checkstyle/meta/checks/naming/RecordComponentNameCheck.xml
+++ b/src/main/resources/com/puppycrawl/tools/checkstyle/meta/checks/naming/RecordComponentNameCheck.xml
@@ -1,0 +1,22 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<checkstyle-metadata>
+   <module>
+      <check fully-qualified-name="com.puppycrawl.tools.checkstyle.checks.naming.RecordComponentNameCheck"
+             name="RecordComponentName"
+             parent="com.puppycrawl.tools.checkstyle.TreeWalker">
+         <description>&lt;p&gt;
+ Checks that record component names conform to a specified pattern.
+ &lt;/p&gt;</description>
+         <properties>
+            <property default-value="^[a-z][a-zA-Z0-9]*$"
+                      name="format"
+                      type="java.util.regex.Pattern">
+               <description>Specifies valid identifiers.</description>
+            </property>
+         </properties>
+         <message-keys>
+            <message-key key="name.invalidPattern"/>
+         </message-keys>
+      </check>
+   </module>
+</checkstyle-metadata>

--- a/src/main/resources/google_checks.xml
+++ b/src/main/resources/google_checks.xml
@@ -214,6 +214,11 @@
       <message key="name.invalidPattern"
              value="Class type name ''{0}'' must match pattern ''{1}''."/>
     </module>
+    <module name="RecordComponentName">
+      <property name="format" value="^[a-z]([a-z0-9][a-zA-Z0-9]*)?$"/>
+      <message key="name.invalidPattern"
+               value="Record component name ''{0}'' must match pattern ''{1}''."/>
+    </module>
     <module name="RecordTypeParameterName">
       <property name="format" value="(^[A-Z][0-9]?)$|([A-Z][a-zA-Z0-9]*[T]$)"/>
       <message key="name.invalidPattern"

--- a/src/test/java/com/puppycrawl/tools/checkstyle/checks/naming/RecordComponentNameCheckTest.java
+++ b/src/test/java/com/puppycrawl/tools/checkstyle/checks/naming/RecordComponentNameCheckTest.java
@@ -1,0 +1,94 @@
+////////////////////////////////////////////////////////////////////////////////
+// checkstyle: Checks Java source code for adherence to a set of rules.
+// Copyright (C) 2001-2021 the original author or authors.
+//
+// This library is free software; you can redistribute it and/or
+// modify it under the terms of the GNU Lesser General Public
+// License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+//
+// This library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+// Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public
+// License along with this library; if not, write to the Free Software
+// Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
+////////////////////////////////////////////////////////////////////////////////
+
+package com.puppycrawl.tools.checkstyle.checks.naming;
+
+import static com.puppycrawl.tools.checkstyle.checks.naming.AbstractNameCheck.MSG_INVALID_PATTERN;
+import static org.junit.jupiter.api.Assertions.assertArrayEquals;
+
+import org.junit.jupiter.api.Test;
+
+import com.puppycrawl.tools.checkstyle.AbstractModuleTestSupport;
+import com.puppycrawl.tools.checkstyle.DefaultConfiguration;
+import com.puppycrawl.tools.checkstyle.api.TokenTypes;
+
+public class RecordComponentNameCheckTest extends AbstractModuleTestSupport {
+
+    @Override
+    protected String getPackageLocation() {
+        return "com/puppycrawl/tools/checkstyle/checks/naming/recordcomponentname";
+    }
+
+    @Test
+    public void testGetClassRequiredTokens() {
+        final RecordComponentNameCheck checkObj =
+                new RecordComponentNameCheck();
+        final int[] expected = {TokenTypes.RECORD_COMPONENT_DEF};
+        assertArrayEquals(expected, checkObj.getRequiredTokens(),
+                "Default required tokens are invalid");
+    }
+
+    @Test
+    public void testRecordDefault()
+            throws Exception {
+        final DefaultConfiguration checkConfig =
+                createModuleConfig(RecordComponentNameCheck.class);
+
+        final String pattern = "^[a-z][a-zA-Z0-9]*$";
+
+        final String[] expected = {
+            "14:34: " + getCheckMessage(MSG_INVALID_PATTERN, "value_123", pattern),
+            "15:15: " + getCheckMessage(MSG_INVALID_PATTERN, "Values", pattern),
+            "18:35: " + getCheckMessage(MSG_INVALID_PATTERN, "_value123", pattern),
+            "19:9: " + getCheckMessage(MSG_INVALID_PATTERN, "$age", pattern),
+        };
+        verify(checkConfig,
+                getNonCompilablePath("InputRecordComponentNameDefault.java"), expected);
+    }
+
+    @Test
+    public void testClassFooName()
+            throws Exception {
+        final DefaultConfiguration checkConfig =
+                createModuleConfig(RecordComponentNameCheck.class);
+        checkConfig.addAttribute("format", "^[a-z0-9]+$");
+
+        final String pattern = "^[a-z0-9]+$";
+
+        final String[] expected = {
+            "15:36: " + getCheckMessage(MSG_INVALID_PATTERN, "value_123", pattern),
+            "16:15: " + getCheckMessage(MSG_INVALID_PATTERN, "Values", pattern),
+            "19:37: " + getCheckMessage(MSG_INVALID_PATTERN, "V", pattern),
+            "20:9: " + getCheckMessage(MSG_INVALID_PATTERN, "$age", pattern),
+        };
+        verify(checkConfig,
+                getNonCompilablePath("InputRecordComponentNameLowercase.java"), expected);
+    }
+
+    @Test
+    public void testGetAcceptableTokens() {
+        final RecordComponentNameCheck typeParameterNameCheckObj =
+                new RecordComponentNameCheck();
+        final int[] actual = typeParameterNameCheckObj.getAcceptableTokens();
+        final int[] expected = {
+            TokenTypes.RECORD_COMPONENT_DEF,
+        };
+        assertArrayEquals(expected, actual, "Default acceptable tokens are invalid");
+    }
+}

--- a/src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/naming/recordcomponentname/InputRecordComponentNameDefault.java
+++ b/src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/naming/recordcomponentname/InputRecordComponentNameDefault.java
@@ -1,0 +1,20 @@
+//non-compiled with javac: Compilable with Java14
+package com.puppycrawl.tools.checkstyle.checks.naming.recordcomponentname;
+
+import java.io.Serializable;
+import java.util.LinkedHashMap;
+
+import org.w3c.dom.Node;
+
+/* Config: default
+ */
+public record InputRecordComponentNameDefault(int x, String str, Double myValue1, int i1) { // ok
+}
+
+record DefaultFirstRecord(String value_123, // violation
+    String... Values) { // violation
+}
+
+record DefaultSecondRecord(String _value123, // violation
+    int $age) { // violation
+}

--- a/src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/naming/recordcomponentname/InputRecordComponentNameLowercase.java
+++ b/src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/naming/recordcomponentname/InputRecordComponentNameLowercase.java
@@ -1,0 +1,21 @@
+//non-compiled with javac: Compilable with Java14
+package com.puppycrawl.tools.checkstyle.checks.naming.recordcomponentname;
+
+import java.io.Serializable;
+import java.util.LinkedHashMap;
+
+import org.w3c.dom.Node;
+
+/* Config:
+ * pattern = "^[a-z\\d]+$"
+ */
+public record InputRecordComponentNameLowercase(Integer x, String str, Double val123) { // ok
+}
+
+record LowercaseFirstRecord(String value_123, // violation
+    String... Values) { // violation
+}
+
+record LowercaseSecondRecord(String V, // violation
+    int $age) { // violation
+}

--- a/src/xdocs/checks.xml
+++ b/src/xdocs/checks.xml
@@ -759,6 +759,14 @@
           </tr>
           <tr>
             <td>
+              <a href="config_naming.html#RecordComponentName">RecordComponentName</a>
+            </td>
+            <td>
+              Checks that record component names conform to a specified pattern.
+            </td>
+          </tr>
+          <tr>
+            <td>
               <a href="config_naming.html#RecordTypeParameterName">RecordTypeParameterName</a>
             </td>
             <td>

--- a/src/xdocs/config_naming.xml
+++ b/src/xdocs/config_naming.xml
@@ -2684,6 +2684,110 @@ class MyClass {
       </subsection>
     </section>
 
+    <section name="RecordComponentName">
+      <p>Since Checkstyle 8.40</p>
+      <subsection name="Description" id="RecordComponentName_Description">
+        <p>Checks that record component names conform to a specified pattern.</p>
+      </subsection>
+
+      <subsection name="Properties" id="RecordComponentName_Properties">
+        <div class="wrapper">
+          <table>
+            <tr>
+              <th>name</th>
+              <th>description</th>
+              <th>type</th>
+              <th>default value</th>
+              <th>since</th>
+            </tr>
+            <tr>
+              <td>format</td>
+              <td>Specifies valid identifiers.</td>
+              <td>
+                <a href="property_types.html#Pattern">Pattern</a>
+              </td>
+              <td>
+                <code>"^[a-z][a-zA-Z0-9]*$"</code>
+              </td>
+              <td>8.40</td>
+            </tr>
+          </table>
+        </div>
+      </subsection>
+
+      <subsection name="Examples" id="RecordComponentName_Examples">
+        <p>
+          To configure the check:
+        </p>
+        <source>
+&lt;module name=&quot;RecordComponentName&quot;/&gt;
+        </source>
+        <p>Example:</p>
+        <source>
+record MyRecord1(String value, int otherComponentName) {} // OK
+record MyRecord2(String... Values) {} // violation, the record component name
+                                    // should match the regular expression "^[a-z][a-zA-Z0-9]*$"
+record MyRecord3(double my_number) {} // violation, the record component name
+                                    // should match the regular expression "^[a-z][a-zA-Z0-9]*$"
+        </source>
+        <p>
+          An example of how to configure the check for names that are only letters in lowercase:
+        </p>
+        <p>Configuration:</p>
+        <source>
+&lt;module name=&quot;RecordComponentName&quot;&gt;
+    &lt;property name="format" value="^[a-z]+$"/&gt;
+&lt;/module&gt;
+        </source>
+        <p>Example:</p>
+        <source>
+record MyRecord1(String value, int other) {} // OK
+record MyRecord2(String... strings) {} // OK
+record MyRecord3(double myNumber) {} // violation, the record component name
+                             // should match the regular expression "^[a-z]+$"
+        </source>
+      </subsection>
+
+      <subsection name="Example of Usage" id="RecordComponentName_Example_of_Usage">
+        <ul>
+          <li>
+            <a href="https://github.com/search?q=path%3Asrc%2Fmain%2Fresources+filename%3Agoogle_checks.xml+repo%3Acheckstyle%2Fcheckstyle+RecordComponentName">
+              Google Style
+            </a>
+          </li>
+          <li>
+            <a href="https://github.com/search?q=path%3Aconfig+filename%3Acheckstyle_checks.xml+repo%3Acheckstyle%2Fcheckstyle+RecordComponentName">
+              Checkstyle Style
+            </a>
+          </li>
+        </ul>
+      </subsection>
+
+      <subsection name="Violation Messages" id="RecordComponentName_Violation_Messages">
+        <ul>
+          <li>
+            <a href="https://github.com/search?q=path%3Asrc%2Fmain%2Fresources%2Fcom%2Fpuppycrawl%2Ftools%2Fcheckstyle%2Fchecks%2Fnaming+filename%3Amessages*.properties+repo%3Acheckstyle%2Fcheckstyle+%22name.invalidPattern%22">
+              name.invalidPattern
+            </a>
+          </li>
+        </ul>
+        <p>
+          All messages can be customized if the default message doesn't suit you.
+          Please <a href="config.html#Custom_messages">see the documentation</a> to learn how to.
+        </p>
+      </subsection>
+
+      <subsection name="Package" id="RecordComponentName_Package">
+        <p>com.puppycrawl.tools.checkstyle.checks.naming</p>
+      </subsection>
+
+      <subsection name="Parent Module" id="RecordComponentName_Parent_Module">
+        <p>
+          <a href="config.html#TreeWalker">TreeWalker</a>
+        </p>
+      </subsection>
+    </section>
+
     <section name="RecordTypeParameterName">
       <p>Since Checkstyle 8.36</p>
       <subsection name="Description" id="RecordTypeParameterName_Description">

--- a/src/xdocs/google_style.xml
+++ b/src/xdocs/google_style.xml
@@ -1660,6 +1660,13 @@
                   </span>
                   <a href="config_naming.html#LambdaParameterName">
                     LambdaParameterName</a>
+                  <br/>
+                  <br/>
+                  <span class="wrapper inline">
+                    <img src="images/ok_green.png" alt=""/>
+                  </span>
+                  <a href="config_naming.html#RecordComponentName">
+                    RecordComponentName</a>
                 </td>
                 <td>
                   <a href="https://github.com/search?q=path%3Asrc%2Fmain%2Fresources+filename%3Agoogle_checks.xml+repo%3Acheckstyle%2Fcheckstyle+ParameterName">
@@ -1680,6 +1687,12 @@
                     config</a>
                   <br />
                   <a href="https://github.com/checkstyle/checkstyle/blob/master/src/it/java/com/google/checkstyle/test/chapter5naming/rule526parameternames/LambdaParameterNameTest.java">
+                    test</a>
+                  <br/>
+                  <a href="https://github.com/search?q=path%3Asrc%2Fmain%2Fresources+filename%3Agoogle_checks.xml+repo%3Acheckstyle%2Fcheckstyle+RecordComponentName">
+                    config</a>
+                  <br />
+                  <a href="https://github.com/checkstyle/checkstyle/blob/master/src/it/java/com/google/checkstyle/test/chapter5naming/rule526parameternames/RecordComponentNameTest.java">
                     test</a>
                 </td>
               </tr>

